Orleans/src/Microsoft.Orleans.Host.Abstractions/StorageProvidersFeature.cs

```diff
@@ -0,0 +1,51 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.Orleans.Host.Abstractions
+{
+    // this would go into the (optional) storage providers DLL.
+    public interface IStorageProvidersBuilder
+    {
+        ISiloBuilder Silo { get; }
+        void AddProvider(IStorageProvider provider);
+    }
+
+    public static class SiloStorageProviderExtensions
+    {
+        public static void AddStorageProviders(this IServiceCollection services)
+        {
+            services.AddSingleton<IStorageProvidersBuilder, StorageProvidersBuilder>();
+        }
+        public static void ConfigureStorageProviders(this ISiloBuilder silo, Action<IStorageProvidersBuilder> configureMethod)
+        {
+            var factory = silo.ApplicationServices.GetRequiredService<IStorageProvidersBuilder>();
+            configureMethod(factory);
```

Comment:
not sure I follow. This is not client code. The programming API for the user does not use the service locator pattern, it's all abstracted away by these configuration methods, in the same way as Asp.NET hides these too. If we change the signature to what you suggested, it means the user will be in charge of resolving the `IStorageProvidersBuilder` himself and then pass it in (probably by putting the burden on him to use the service locator pattern